### DISCUSSION TITLE
[feature request] Support nested (multi-part) Namespaces / Schemas

### DISCUSSION THREAD

**Discussion by c-thiel** [unitycatalog/unitycatalog#215] [Ideas]
**Is your feature request related to a problem? Please describe.**
The Iceberg-REST specification explicitly allows arbitrary nested namespaces modeled as an array of string.
Currently UC only supports two levels: <Database>.<Schema>. 

Many query engines such as Spark, Dremio and Pyiceberg support nested Namespaces out of the box. Nested Namespaces are an important feature to be able to structure Data more fine-granular, without utilizing artificial separators such as `_` in Database or Namespace names. In a Data Mesh scenario, a typical structure could look like:
`<Business-Unit>.<Layer (dev / test / prod)>.<Data Product or Use Case>.<Layer (bronze / silver / gold).<Table>`

Today many companies need to augment this structure using composed names, which is a bad practice.

**Describe the solution you would like**
It would be great if UC would support multi-part Namespaces / Schemas

**Comment by MrPowers** [comment]
@c-thiel - thanks for raising this fascinating idea.

As you mentioned, UC instances are currently organized in a three-level hierarchy as `catalog_name.schema_name.data_asset_name`.

You're saying that this three-level namespace forces you to do something like this: `cool_catalog.dev_accounting_bronze.salary` and you would rather have this: `cool_catalog.dev.accounting.bronze.salary`.

I could see this going either way.  The three-level namespace requires consistency and simplicity.  In this example, perhaps the organization could use `dev.accounting_bronze.salary`.  But I could also see how a namespace with arbitrary length could be nice.

**Reply by creechy** [reply]
Is there more functionality that opens up by supporting multi-level namespaces? Or just that it feels less artificial than using an arbitrary character like underscore, which in my experience organizations are generally fine with.

The only thing I can imagine functionally would be to (eventually) manage permissions at each level, which could be pretty challenging, and begs the question if the effort is worth it?

Also, many engines like Trino are used to the three part naming, introducing additional levels could pose issues for them, and may lead to odd syntax to address them `SELECT * from catalog."dev.accounting.bronze".table`.
